HARK/ConsumptionSaving/ConsPortfolioFrameModel.py

```diff
@@ -0,0 +1,286 @@
+"""
+This file contains classes and functions for representing,
+solving, and simulating agents who must allocate their resources
+among consumption, saving in a risk-free asset (with a low return),
+and saving in a risky asset (with higher average return).
+
+This file also demonstrates a "frame" model architecture.
+"""
+import numpy as np
+from scipy.optimize import minimize_scalar
+from copy import deepcopy
+from HARK import NullFunc, Frame, FrameAgentType  # Basic HARK features
+from HARK.ConsumptionSaving.ConsIndShockModel import (
+    IndShockConsumerType,  # PortfolioConsumerType inherits from it
+    utility,  # CRRA utility function
+    utility_inv,  # Inverse CRRA utility function
+    utilityP,  # CRRA marginal utility function
+    utility_invP,  # Derivative of inverse CRRA utility function
+    utilityP_inv,  # Inverse CRRA marginal utility function
+    init_idiosyncratic_shocks,  # Baseline dictionary to build on
+)
+from HARK.ConsumptionSaving.ConsRiskyAssetModel import (
+    RiskyAssetConsumerType
+)
+from HARK.ConsumptionSaving.ConsPortfolioModel import (
+    init_portfolio,
+    solveConsPortfolio,
+    PortfolioConsumerType,
+    PortfolioSolution
+)
+
+from HARK.distribution import combine_indep_dstns, add_discrete_outcome_constant_mean
+from HARK.distribution import (
+    IndexDistribution,
+    Lognormal,
+    MeanOneLogNormal,
+    Bernoulli  # Random draws for simulating agents
+)
+from HARK.interpolation import (
+    
+    LinearInterp,  # Piecewise linear interpolation
+    CubicInterp,  # Piecewise cubic interpolation
+    LinearInterpOnInterp1D,  # Interpolator over 1D interpolations
+    BilinearInterp,  # 2D interpolator
+    ConstantFunction,  # Interpolator-like class that returns constant value
+    IdentityFunction,  # Interpolator-like class that returns one of its arguments
+    ValueFuncCRRA,
+    MargValueFuncCRRA,
+    MargMargValueFuncCRRA
+)
+
+class PortfolioConsumerFrameType(FrameAgentType, PortfolioConsumerType):
+    """
+    A consumer type with a portfolio choice, using Frame architecture.
+
+    A subclass of PortfolioConsumerType for now.
+    This is mainly to keep the _solver_ logic intact.
+    """
+
+    def __init__(self, **kwds):
+        params = init_portfolio.copy()
+        params.update(kwds)
+        kwds = params
+
+        # Initialize a basic consumer type
+        PortfolioConsumerType.__init__(
+            self, **kwds
+        )
+
+        # Set the solver for the portfolio model, and update various constructed attributes
+        self.solve_one_period = solveConsPortfolio
+        self.update()
+
+        ## TODO: Should be defined in the configuration.
+        self.aggs = {'PermShkAggNow' : None, 'PlvlAgg' : None} # aggregate values
```

Comment:
I like what you have done here, fully specifying states, shocks, controls and aggregates.

When creating new models, a frustration of mine has been the inconsistency with which that specification is done across different HARK agent types. Its sometimes done, sometimes not, sometimes inherited so one has to go hunting for the original class' attributes, sometimes fully specified.

It is further complicated by two different objects  `shock_vars` and `shock_vars_` ([see e.g.](https://github.com/econ-ark/HARK/blob/10fe1efba5d6f558fc47433325ac987acfc09077/HARK/ConsumptionSaving/ConsIndShockModel.py#L1583)) with no indication of what each of them does. I think I have seen this for states too.

Reply:
What I would like to do ultimately is move all of this information into the frame definitions.
Then have the constructor code build these dictionaries based on those frames.
Ultimately, I want all model information in the frames dictionary, and nothing baked into a model class.
This is sort of an intermediary implementation.